### PR TITLE
IBX-3445: Softened `TargetContentValidator` content check

### PR DESCRIPTION
**Pull request by barw4**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-3445](https://issues.ibexa.co/browse/IBX-3445)
| **Type**                                   | improvement
| **Target Ibexa version** | `v3.3`
| **BC breaks**                          | no

Mentioned PR will allow to update content that has relation not available to a current user. 

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x]  Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/engineering-team`).

### CODE REVIEW

File: src/lib/Repository/Validator/TargetContentValidator.php [hunk 1]
```diff
@@ -51,13 +45,11 @@ public function validate(int $value, array $allowedContentTypes = []): ?Validati
                     'targetContentId'
                 );
             }
-        } catch (NotFoundException | UnauthorizedException $e) {
+        } catch (NotFoundException $e) {
             return new ValidationError(
-                'Content with identifier %contentId% is not a valid relation target',
```
Comment (adamwojs) [inline src/lib/Repository/Validator/TargetContentValidator.php:56]: What is the reason behind this change ?

File: src/lib/Repository/Validator/TargetContentValidator.php [hunk 2]
```diff
@@ -21,25 +19,21 @@
...
 
     public function validate(int $value, array $allowedContentTypes = []): ?ValidationError
     {
         try {
-            $contentInfo = $this->contentService->loadContentInfo($value);
-            $contentType = $this->contentTypeService->loadContentType($contentInfo->contentTypeId);
+            $contentInfo = $this->repository->sudo(static function (Repository $repository) use ($value) {
+                return $repository->getContentService()->loadContentInfo($value);
+            });
+            $contentType = $this->repository->getContentTypeService()->loadContentType($contentInfo->contentTypeId);
```
Comment (adamwojs) [inline src/lib/Repository/Validator/TargetContentValidator.php]: 1) return type is missing in callback 
2) Please consider using \eZ\Publish\SPI\Persistence\Content\Handler::loadContentInfo and \eZ\Publish\SPI\Persistence\Content\Type\Handler::load instead of `sudo`

**Review by adamwojs — COMMENTED** [review]
Related discussion: https://github.com/ezsystems/ezpublish-kernel/pull/3131#discussion_r858570959

File: src/lib/Repository/Validator/TargetContentValidator.php (hunk 1 above)
Comment (barw4) [reply src/lib/Repository/Validator/TargetContentValidator.php:56]: That was @glye idea, please see the internal comment on the JIRA ticket.

File: src/lib/Repository/Validator/TargetContentValidator.php [hunk 3]
```diff
@@ -21,25 +19,21 @@
...
     }
 
     public function validate(int $value, array $allowedContentTypes = []): ?ValidationError
     {
         try {
-            $contentInfo = $this->contentService->loadContentInfo($value);
-            $contentType = $this->contentTypeService->loadContentType($contentInfo->contentTypeId);
+            $contentInfo = $this->repository->sudo(static function (Repository $repository) use ($value) {
+                return $repository->getContentService()->loadContentInfo($value);
+            });
```
Comment (alongosz) [inline src/lib/Repository/Validator/TargetContentValidator.php]: Please avoid injecting Repository to fetch services at runtime. This is allowed for integration tests only. For production code it's something that I plan to get rid of at some point. Repository was an attempt to wrap parts of eZ Publish architecture, in effect creating a container within a Symfony container. The same can be achieved on a SF Framework level without intermediate classes. There's also underlying performance issue there. For some reason we have at runtime 2 instances of `Repository` which occupy memory. Moreover fetching services via getter methods is slightly slower than when injected into constructor and you might end-up with a service obtained without proper SA context.

If you want `sudo`, there are 2 things to inject ATM (needs to be changed as well): `PermissionResolver` and `Repository`, while the latter one can be used only as a formal argument of `PermissionResolver::sudo(callback, Repository)` method call.
So it should be rather something like this:
```php
            $contentInfo = $this->permissionResolver->sudo(
                function (Repository $repository) use ($value): ContentInfo {
                    return $this->contentService->loadContentInfo($value);
                },
                $this->repository
            );
```
:exclamation:  BUT before applying please read other comments, in this case we don't need to inject API and use `sudo`

File: src/lib/Repository/Validator/TargetContentValidator.php [hunk 4]
```diff
@@ -21,25 +19,21 @@
...
-    public function __construct(
-        ContentService $contentService,
-        ContentTypeService $contentTypeService
-    ) {
-        $this->contentService = $contentService;
-        $this->contentTypeService = $contentTypeService;
+    public function __construct(Repository $repository)
+    {
+        $this->repository = $repository;
     }
```
Comment (alongosz) [inline src/lib/Repository/Validator/TargetContentValidator.php]: This is not necessary, let's load lighter Persistence Layer.
```php
use eZ\Publish\SPI\Persistence\Content;
// ...
    public function __construct(
        Content\Handler $contentHandler,
        Content\Type\Handler $contentTypeHandler
    ) {
        $this->contentHandler = $contentHandler;
        $this->contentTypeHandler = $contentTypeHandler;
    }
```

File: src/lib/Repository/Validator/TargetContentValidator.php [hunk 5]
```diff
@@ -21,25 +19,21 @@
...
     {
         try {
-            $contentInfo = $this->contentService->loadContentInfo($value);
-            $contentType = $this->contentTypeService->loadContentType($contentInfo->contentTypeId);
+            $contentInfo = $this->repository->sudo(static function (Repository $repository) use ($value) {
+                return $repository->getContentService()->loadContentInfo($value);
+            });
+            $contentType = $this->repository->getContentTypeService()->loadContentType($contentInfo->contentTypeId);
 
             if (!empty($allowedContentTypes) && !in_array($contentType->identifier, $allowedContentTypes, true)) {
```
Comment (alongosz) [inline src/lib/Repository/Validator/TargetContentValidator.php]: So following the previous comment:
```php
            $contentInfo = $this->contentHandler->loadContentInfo($value);
            $contentType = $this->contentTypeHandler->load($contentInfo->contentTypeId);

            if (
                !empty($allowedContentTypes) &&
                !in_array($contentType->identifier, $allowedContentTypes, true)
            ) {
```

**Review by alongosz — CHANGES_REQUESTED** [review]

File: src/lib/Repository/Validator/TargetContentValidator.php (hunk 3 above)
Comment (barw4) [reply src/lib/Repository/Validator/TargetContentValidator.php]: Thanks for the explanation

**Comment by barw4** [comment]
Not sure about the CS check, both files were cs-fixed

**Comment by Steveb-p** [comment]
> Not sure about the CS check, both files were cs-fixed

It's highly likely a php-cs-fixer update "uncovered" some new issues. Re-run `composer update` & `composer fix-cs`.

File: src/lib/Repository/Validator/TargetContentValidator.php (hunk 1 above)
Comment (Steveb-p) [reply src/lib/Repository/Validator/TargetContentValidator.php:56]: To be frank saying that some Content ID does not exist (because that's what will actually be happening) is hardly a security issue. It's not like we're saying what usernames there are. Even if an ID is leaked, user is still not capable of reaching it. And we're depriving ourselves of a rather important debugging information :disappointed: 

Like in many other places, we're losing critical information, and when dealing with any bugs related to this code we will need to manually expose the ID.

So for me it's either this shouldn't be changed, or we need a follow up sometime to include relevant debugging information. (Adding internal information to an exception is not uncommon)

**Review by Steveb-p — APPROVED** [review]

**Review by alongosz — COMMENTED** [review]
The CS issue is related to the changes, due to outdated fixer. Proceed as Steve suggested above please.

**Comment by barw4** [comment]
@alongosz @Steveb-p CS was fixed and I also reverted the change with removing content identifier.

**Review by alongosz — APPROVED** [review]

**Review by Steveb-p — APPROVED** [review]

**Review by konradoboza — APPROVED** [review]

**Review by micszo — APPROVED** [review]
QA Approved on Ibexa Commerce 3.3.27-dev.
